Add support for blackboard types

bjoernschroeder opened the pull request:
Wrike: https://www.wrike.com/open.htm?id=940589144

bjoernschroeder committed:
- check if set blackboard value matches type in pml
- throw custom exception in blackboard get

bjoernschroeder committed:
set default values

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,60 @@
+    {
+    }
+
+    std::string what()
```
`what()` is supposed to have the signature: https://en.cppreference.com/w/cpp/error/exception/what & derived classes are expected to override it, that way users need only catch `std::exception` & polymorphism ensures the right `what()` method is called. Also, the reason it only returns a const char* is probably because the exception object should itself not throw any more exceptions. Here std::string, std::stringstream etc. can throw (mainly due to dynamic memory allocation)

So I think we can have 2 methods: 
1. override what() as per https://en.cppreference.com/w/cpp/error/exception/what which returns a simple msg (static const char* string so that no dynamic memory allocation is needed) &
2. have another method which returns this more detailed output.

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,60 @@
+    std::string _typeDefinedInPml;
+};
+
+class BadBlackboardAccessType : public std::exception
```
I think we can have a single exception instead of 2? `BlackboardTypeMismatch`?

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +68,100 @@ class BlackboardImpl
+                std::string typeString = entry[Strings::stateType].as<std::string>();
+                if (typeString == "int64_t") {
+                    set<int64_t>(key, defaultValue.as<int64_t>());
+                } else if (typeString == "bool") {
```
can define these strings as `constexpr const char*` constants. And can follow ROS convention i.e. int64_t -> int64

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +68,100 @@ class BlackboardImpl
+    }
+
+    template <class T>
+    std::string getNameFromType() const
```
seems like there is a simple way to do the conversion from type to name without having an if else ladder. Bonus: supports implicit conversions like int32_t to int64_t etc.
But converting from std::string to type is not so easy.
Also can add more types easily although I can't think of too many useful types that can be added. Adding ROS types will require a little bit of tweaking, but shouldn't be too difficult

```
struct Converter {
    static constexpr const char* typeNames[] = {"bool", "int64", "double", "std::string", "std::any"};
    using Types = std::variant<bool, int64_t, double, std::string, std::any>;

    // convert from type to name as string
    template <class T>
    static constexpr const char* typeName() {
        return typeNames[Types{T{}}.index()];
    }

    // set the value given the type name as a string
    static std::any setDefaultValue(const std::string& tpname, const std::string& value) {
        static constexpr std::size_t numTypes = sizeof(typeNames) / sizeof(const char*);
        for (std::size_t i = 0; i < numTypes; ++i) {
            if (tpname == typeNames[i]) {
                auto value = makeAnyFromIndex(Types{}, std::make_index_sequence<numTypes>{}, i, value);
                // store value on the blackboard
            }
        }
        return {};
    }

    template <class T>
    struct YamlAs {
        static T as(const std::string& value) {
            // call Yaml's as<T>(value) method
        }
    };

    template <class T, std::size_t INDEX, template <class> class Parser = YamlAs>
    struct makeAnyIfEqual {
        static std::any make(std::size_t index, const std::string& value) {
            return index == INDEX ? std::any{Parser<T>::as(value)} : std::any{};
        }
    };

    template <class... Ts, std::size_t... Is>
    static std::any makeAnyFromIndex(std::variant<Ts...>, std::index_sequence<Is...>, std::size_t index, const std::string& value) {
        std::any typeValues[] = { makeAnyIfEqual<Ts, Is>::make(index, value)... };
        return typeValues[index];
    }
};
```

bjoernschroeder committed:
- use only a single exception for blackboard type mismatches
- add method for more detailed exception output
- use converter to set default blackboard values

bjoernschroeder committed:
- set default values with the correct type
- pass blackboard parameters with the correct type
- get type of value without if-else ladder

bjoernschroeder committed:
- set correct types for blackboard values
- fix tests
- access bb values with the correct type

bjoernschroeder committed:
store std::variant instead of std::any in blackboard

bjoernschroeder committed:
update blackboard value access

bjoernschroeder committed:
move std::variant type to Types.h, remove old code, remove unused includes

veerajsk commented on alica_engine/include/engine/Types.h:
```diff
@@ -62,6 +64,7 @@ using VariableGrp = std::vector<const Variable*>;
 using ParameterMap = std::unordered_map<std::string, Parameter*>;
 using AgentStatePair = std::pair<AgentId, const State*>;
 using TransitionConditionCallback = std::function<bool(const Blackboard*, const RunningPlan*, const IAlicaWorldModel*)>;
+using BlackboardValue = std::variant<bool, int64_t, double, std::string, std::any>;
```
`BlackboardValue` => `BlackboardValueType`

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,46 @@
+
+    const char* what() const throw() override { return "The type of the blackboard value does not match the type in the pml file."; }
+
+    std::string detailedError()
```
`std::string detailedError()` => `std::string detailedError() const`

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,46 @@
+    }
+
+private:
+    std::string _inputType;
```
`const std::string` for both

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -1,15 +1,46 @@
+    {
+    }
+
+    const char* what() const throw() override { return "The type of the blackboard value does not match the type in the pml file."; }
```
use `noexcept` instead of `throw()` since the latter is deprecated

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+    void set(const std::string& key, const T& value)
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        if (entry[Strings::key].as<std::string>() != key) {
```
won't this throw an exception because an empty `Yaml::Node()` is returned?

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+    {
+        static BlackboardValue make(std::size_t index, const YAML::Node& value)
+        {
+            if (index == INDEX) {
```
this is not required right? Can just do
`return index == INDEX ? BlackboardValue{Parser<T>::as(value)} : BlackboardValue{std::any{}};`

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+    };
+
+    template <class... Ts, std::size_t... Is>
+    static void setBlackboardValue(
```
is this really required? Can't we just store the variant returned by `makeVariantFromIndex()` in the `vals` `unordered_map`?

veerajsk commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+
+    // YAML doesnt support std::any, needs a separate list of types for iteration
+    static constexpr const char* typeNamesYAML[] = {"bool", "int64", "double", "std::string"};
+    using TypesYAML = std::variant<bool, int64_t, double, std::string>;
```
Can't we get rid of this as well? The type of the default value should be one of the strings in `typeNames[]` (apart from `std::any`). If it is not, we simply reject setting the value (for now).

bjoernschroeder committed:
add getAnyAs function, simplify blackboard code
